### ISSUE TITLE
Support zsh abbreviations, bindkeys, and history expansion

### ISSUE THREAD

**Issue by juliogarciag** [warpdotdev/Warp#537]
### Describe the bug

[Fish abbreviations](https://fishshell.com/docs/current/cmds/abbr.html) work just fine, but zsh abbreviations coming from the [zsh-abbr plugin](https://github.com/olets/zsh-abbr) don't work at all. When I press `space`, nothing happens.

### To Reproduce

Steps to reproduce:

1. Use zsh
2. Install this plugin: https://github.com/olets/zsh-abbr
3. Try to create an abbreviation: `abbr gs='git status'`
4. Try to use the abbreviation by typing: `gs` and then `space`

### Expected behaviour

```shell
The abbreviation should be expanded
```

**Comment by juliogarciag** [comment]
After taking a deeper look, it seems `bindkey` doesn't work yet. Is that something scheduled to be tackled soon?

**Comment by elviskahoro** [comment]
@juliogarciag CC'ing @charlespierce as he's working on custom keybindings and this falls under that umbrella!

**Comment by elviskahoro** [comment]
are there plans to do left option as ESC+ instead of Meta? I use ESC+. which is insert-last-word on zsh all the time and I just get ≥ regardless of whether Left Option Key is Meta is enabled
$_ is my current workaround. Also might be for bugs but using actual ESC+. doesn't seem to work, so I'm guessing it's just a feature of Zsh Line Editor that isn't implemented at all

**Comment by linux-china** [comment]
Same problem to me, and bindkey does not work for zsh autosuggestions. 

```shell
# zsh auto suggestions
source /usr/local/share/zsh-autosuggestions/zsh-autosuggestions.zsh
bindkey '^k' autosuggest-accept
bindkey '^j' autosuggest-execute
```

**Comment by elviskahoro** [comment]
From:
- #248 

### Describe the bug

Try putting the below in the ~/.zshrc
```
bindkey "^j" history-search-forward
```

### To Reproduce

_No response_

### Expected behaviour

_No response_

### Screenshots

https://share.getcloudapp.com/7KuE7w6K

### Operating System

_No response_

### OS Version

_No response_

### Additional context
No other bindkey seems to present an issue

**Comment by elviskahoro** [comment]
From @hardselius via:
- #589 

### Describe the bug

I have the following snippet in my `.zshrc`

```shell
resume() {
  fg
  zle push-input
  BUFFER=""
  zle accept-line
}
zle -N resume
bindkey "^Z" resume
```

It serves as a "toggle" for programs running in the terminal, e.g vim.

### To Reproduce

Steps to reproduce:
1. Add the snippet to .zshrc
2. Open vim
3. Press ^Z to move vim to the background
4. Press ^Z again and nothing happens.


### Expected behaviour

In this case, I expect `^Z` to bring my process back to the foreground, instead I have to manually run `fg`.

### Screenshots

_No response_

### Operating System

MacOS

### OS Version

12.1

### Additional context

_No response_

**Comment by psbanka** [comment]
It's very convenient to be able to use `META-b` or `META-f` to go forward and backward a word at a time in the terminal. Also, `META-.` to retrieve the last word used in the previous command. Would love to be able to have the option key bound to META rather than needing to use ESC.

**Comment by elviskahoro** [comment]
@psbanka is "Left option key is meta" on for you?
Mac Menu > File
<img width="364" alt="Screen Shot 2022-01-24 at 2 46 46 PM" src="https://user-images.githubusercontent.com/29553206/150853570-f8c82e11-2983-4427-a343-4ab954b38a87.png">

**Comment by psbanka** [comment]
Oh, wow, that's great, @elviskahoro , thank you! The last place I expected to find that was under the `File >` menu.

**Comment by elviskahoro** [comment]
@psbanka There's room to improve our discoverability. Where did you look for it? I'll see if we can get it added there, you're definitely not the first to ask!

**Comment by elviskahoro** [comment]
Moving to:
- #537

### Describe the bug

Zsh's history expansion does not work in Warp (v0.2022.02.21.08.55.stable_00). It seems to work correctly in the stock Terminal app.

### To Reproduce

1. Type a history designator, e.g. `!999`
2. Press <kbd>tab</kbd>

### Expected behaviour

Element 999 (shown in `history`) should be inserted in-place.

### Screenshots

_No response_

### Operating System

MacOS

### OS Version

12.2.1

### Additional context

[Here's my dotfiles](https://github.com/r-thomson/dotfiles/blob/master/zshrc), if it helps

**Comment by denisidoro** [comment]
Sorry for necrobumping this thread, but is there a plan / ETA for supporting bindkeys?

This is a deal-breaker to me, in particular.

**Comment by denisidoro** [comment]
I love the direction Warp is going (eg https://github.com/warpdotdev/Warp/discussions/625) but limiting the user to Warp-native solutions is too strict.

I myself have a plethora of fzf-based-autocompletions (cf https://github.com/warpdotdev/Warp/issues/149) which, in my case, are much more evolved than what Warp currently offers.

I'd really like to use Warp as my daily driver because of its other features (eg blocks) but not being able to do what a vanilla Terminal + zsh shell can offer has prevented me from switching.

**Comment by varenc** [comment]
Lack of support for any zsh keybindings is really a deal breaker for me using Warp. It's also just painfully inelegant.  Zsh has a whole system for managing keybindings, which I'm invested it, and Warp just throws that away.  Even when there's a zsh keybinding defined that doesn't overlap with a Warp binding it still doesn't pass the event along. This also means that little tweaks on default zsh keybindings can't be changed.

To use an example, I [redefine some basic keybindings like this](https://gist.github.com/varenc/e4a22145c484771f254fa20982e2cd7f) so that things like Ctrl-K `kill-line` integrate with the system pasteboard.  Warp just ignores my Ctrl-K keybinding and does the zsh default kill-line, but without my customization the killed line doesn't end up in the system pasteboard.   This uncanny valley of support is quite frustrating.

**Comment by varenc** [comment]
I found a partial workaround: Once in Warp, just run `zsh` again, embedding a zsh session in the already existing zsh session. This results in you losing most all special Warp features, but then all of zsh's features will re-appear.  It sort of just turns Warp into speedy standard terminal emulator like all the others. This doesn't make sense for me but someone who daily drives Warp and occasionally needs some unsupported zsh feature might find this hack useful.

My custom keybinding on Ctrl-K still doesn't work, but other custom bindings do, as well my own history searching method. Could make sense for occasional use. ¯\_(ツ)_/¯

**Comment by elviskahoro** [comment]
> I use the zsh-autosuggestions plugin. It allows you to configure an autosuggest-execute hotkey, which I've configured to Ctrl-P. This works in iTerm2, but no Warp. This makes every auto-suggestion two key presses now, Ctrl-F to "accept autosuggestion" and then Enter to execute it. I'd like a shortcut to do both.

**Comment by robbienohra** [comment]
> I found a partial workaround: Once in Warp, just run `zsh` again, embedding a zsh session in the already existing zsh session. This results in you losing most all special Warp features, but then all of zsh's features will re-appear. It sort of just turns Warp into speedy standard terminal emulator like all the others. This doesn't make sense for me but someone who daily drives Warp and occasionally needs some unsupported zsh feature might find this hack useful.
> 
> My custom keybinding on Ctrl-K still doesn't work, but other custom bindings do, as well my own history searching method. Could make sense for occasional use. ¯_(ツ)_/¯

Well played, sir, lol.

Yeah this is annoying eh, seems like a big miss. I can see how the embedded fuzzy search on command history would reduce my need to use `fzf` for that specific use case. But as you said there is so much more than just fuzzy searching your command history (directories, files, etc.)

Hopefully they can find a solution for this.

**Comment by zheng** [comment]
Most of the comments are about bindkeys, so I'm going to change this title + description to be about bindkeys.

I made https://github.com/warpdotdev/Warp/issues/1514 to keep the original issue description

**Comment by chrisgrieser** [comment]
Either support for `bindkey` or a warp-feature to create bindings to custom commands is indeed dearly missed.

I just want to add, that currently, the Warp docs mention that [you can use `zshrc` to create custom keybindings](https://docs.warp.dev/getting-started/using-warp-with-shells#customize-your-zsh-shell-environment).
> You can use the .zshrc file to customize behavior like: (...) It can also be used to set up key-bindings, and scripts that will automatically execute when a new instance of Zsh is launched.

However, this is not the case as long as there isn't support for `bindkeys` yet, so the docs are a bit misleading here.

**Comment by oiwn** [comment]
> I found a partial workaround: Once in Warp, just run `zsh` again, embedding a zsh session in the already existing zsh session. 

just tried because default fzf bind CTRL+T (fuzzy search files in directory) do not working. Running tmux session also working it drop default warp features and left it as "normal terminal".

**Comment by Gorthog** [comment]
Is this on anyone's radar? This is a deal breaker or me as I use a great plugin to fish that provides me with history + timestamps of command executions.

**Comment by izuzak** [comment]
Just giving this a friendly bump and a 👍  since it's something I'd love in Warp 🙏 🙏

**Comment by squiter** [comment]
I think I have the same problem mentioned by anyone else, but I use bash instead of zsh.
I noticed that Warp ignores the `~/.inputrc` file and the `bind` command from Bash. I think is the same problem of the zsh.

So, I'm just leaving a little bit more context and a friendly thumbs up to this issue. This is the only reason that I didn't drop iTerm2 completely.

**Comment by dannyneira** [comment]
Hey Folks thanks for commenting. As for right now although Warp doesn't support `~/.inputrc` , but as a workaround it does have `~./.warp/keybindings.yaml` which you can learn more about here: https://docs.warp.dev/features/keyboard-shortcuts#custom-keyboard-shortcuts

**Comment by Integralist** [comment]
@dannyneira how would this help in a case like:

```
bindkey '^f' fzf
```

I can't override the default keymap because (as far as I understand the implementation) I have to use a predefined action where as I want to execute a binary (i.e. `fzf`):

```
"<WHAT-GOES-HERE?>": ctrl-f
```

**Comment by dannyneira** [comment]
@Integralist Yeah in this case we're currently not able to bind keyboard shortcuts to binaries or scripts. You're best to go with an alias for quick access to `fzf`. The current implementation of `keybindings.yaml` is only for existing features/actions in Warp. Any updates to this we'll post here!

**Comment by Divide-By-0** [comment]
> Hey Folks thanks for commenting. As for right now although Warp doesn't support `~/.inputrc` or `bindkey`, it does have `~./.warp/keybindings.yaml` which you can use to bind keyboard shortcuts to existing Warp features/actions. Learn more about it here: https://docs.warp.dev/features/keyboard-shortcuts#custom-keyboard-shortcuts
> 
> I want to set realistic expectations that for this feature, we'd have to revamp significantly the way Warp's input editor works so it may be a while before we have any traction. Any updates to this we'll post here!

Can there be some sort of helper script that converts my .zshrc to a slimmed down .zshrc and auto-populates the .keybindings file and whatever else warp uses?

**Comment by Dentrax** [comment]
As fresh user of Warp, I hit this issue. I cleared default key binding of `^r` and put the `bindkey '^f' fzf` in my `.zshrc`. But it does not work. Any ideas how to tackle with this? Otherwise, I'll think about to change my default terminal app...

**Comment by elviskahoro** [comment]
@Dentrax You are correct that it isn't possible to bind a key to an app. We'd have to explicitly build support for this into Warp's Input Editor, since it's native

**Comment by hyperfocus1337** [comment]
This is a shame, I was about to adopt Warp but this is making me reconsider.

I have the following functionality defined in my fish config, which allows me to easily switch directories.

```
function cd_with_fzf
  cd $HOME && cd (fd -t d | fzf --preview="tree -L 1 {}" --bind="space:toggle-preview" --preview-window=:hidden) && echo $PWD && tree -L 2
end

bind \co cd_with_fzf
```

As bind doesn't work, I cannot use it. And there doesn't seem to be way to invoke custom functions with the Warp keybindings (only native keybinds).

Would be happy to see this implemented, I don't necessarily care about `bind` perse, but I just need a way to invoke custom functions using keybindings.

Although it feels like the optimal design should simply support native features that are provided by the underlying shell.

**Comment by rogerkerse** [comment]
I don't see why there is a need to block bindings if in Warp settings no shortcut is conflicting. Having this feature missing is really now allowing me to recommend this terminal to anyone.

**Comment by dannyneira** [comment]
`atuin` isn't compatible with Warp's input, but we've recently added [rich history information](https://docs.warp.dev/features/entry/command-history) which should help some. 

![CleanShot 2023-08-31 at 14 51 27](https://github.com/warpdotdev/Warp/assets/16809145/5f19581d-c252-47e8-a59b-bf9c35e0701c)

Warp also has Command Search which you can use to filter through the history as well.

I hope this helps some with your use cases @fulmicoton

**Comment by ellie** [comment]
> atuin isn't compatible with Warp's input

(hi, Atuin maintainer here)

We're more than happy to hook into Warp to make sure keybindings work, do you have any guidance? If supporting standards is off the table, anyways.

**Comment by tkolleh** [comment]
> `atuin` isn't compatible with Warp's input, but we've recently added [rich history information](https://docs.warp.dev/features/entry/command-history) which should help some.
> 
> ![CleanShot 2023-08-31 at 14 51 27](https://private-user-images.githubusercontent.com/16809145/264783127-5f19581d-c252-47e8-a59b-bf9c35e0701c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI1NDEzNDEsIm5iZiI6MTcwMjU0MTA0MSwicGF0aCI6Ii8xNjgwOTE0NS8yNjQ3ODMxMjctNWYxOTU4MWQtYzI1Mi00N2U4LWE1OWItYmY5YzM1ZTA3MDFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE0VDA4MDQwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlmNWQ2NDIwMDQxMmQ5MDY1ZTgwYTg4ZDdiMmM3MGRhNjE4ZDFmOTYxY2MxMGU1OTdjNzNmZTk0ZmE3OWE3ZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.q0kWD3OAf--6RM6YB75RZ77tqyZvyIWIq6rqNeopK60)
> 
> Warp also has Command Search which you can use to filter through the history as well.
> 
> I hope this helps some with your use cases @fulmicoton

Does warp support keyboard shortcut for history search. I prefer it over the default show history. Adding a up arrow shortcut doesn't work.

<img width="676" alt="Screenshot 2023-12-14" src="https://github.com/warpdotdev/Warp/assets/3095197/4192aa39-0e0b-4c55-8b9e-e2f0658c49de">

**Comment by cowboygneox** [comment]
+1 to having bindkey working.

**Comment by Divide-By-0** [comment]
As a temporary fix, can you just parse such files and manually ignore all bindkey lines?

**Comment by AugustDev** [comment]
This needs to be implemented ASAP as the user experience now is not good. It's pretty embarrassing this issue is open for 2 years.

**Comment by webdev4422** [comment]
+1

**Comment by nikitavoloboev** [comment]
I want to bind a key to a Fish function, currently it fails..

**Comment by nikitavoloboev** [comment]
<img width="340" alt="image" src="https://github.com/warpdotdev/Warp/assets/6391776/8b6183da-c1cd-40ca-bf26-344f5d69aa8d">

i.e. if I put above in fish config, reload shell and trigger `control+g` nothing happens, but it should

**Comment by nikitavoloboev** [comment]
> This is a shame, I was about to adopt Warp but this is making me reconsider.
> 
> I have the following functionality defined in my fish config, which allows me to easily switch directories.
> 
> ```
> function cd_with_fzf
>   cd $HOME && cd (fd -t d | fzf --preview="tree -L 1 {}" --bind="space:toggle-preview" --preview-window=:hidden) && echo $PWD && tree -L 2
> end
> 
> bind \co cd_with_fzf
> ```
> 
> As bind doesn't work, I cannot use it. And there doesn't seem to be way to invoke custom functions with the Warp keybindings (only native keybinds).
> 
> Would be happy to see this implemented, I don't necessarily care about `bind` perse, but I just need a way to invoke custom functions using keybindings.
> 
> Although it feels like the optimal design should simply support native features that are provided by the underlying shell. In this case, first simply support the features of the underlying shell, before adding gimmicky stuff on top.
> 
> Edit: Actually I do care about `bind`, as I would prefer my fish shell configuration to remain cross compatible amongst different terminal emulators. Although adding another custom keybind to execute custom functions inside of Warps configuration would solve the issue usability wise. But best case scenario would be if `bind` simply works out of the box, with the potential of overlapping keybinds amongst `bind` and Warp. Which a user would then simply have to figure out himself. I'm okay with clearing a Warp keybinding to allow for 'room' to use the `bind` configured keybind.

Everything that is said here is spot on.

I hope Warp team actually looks into the issue and supports the underlying shell functionality. 🙏

**Comment by rspears74** [comment]
Just chiming in to say that I would also like to see support for this. I use the zsh `sudo` plugin (double ESC inserts sudo at the beginning of the line), and it doesn't work in Warp.

**Comment by IVIJL** [comment]
This is only thing which blocks me to switch to warp, unable to use my keybinding in .zshrc :(

**Comment by gggiovanny** [comment]
I'm also trying to use fzf, and it's working flawlessly in item2, but not in warp :( a shame, I love warp!

**Comment by DomenicWalther** [comment]
Well, was just configuring my Warp Installation but hit this roadblock, back to iTerm we go

**Comment by dabstractor** [comment]
Total deal-breaker for me. All the cool features amount to nothing if they won't integrate with other tools that have earned their place in the ecosystem. Hopefully you guys can find a resolution.

**Comment by jasek5** [comment]
Is there any ETA ?

**Comment by ellie** [comment]
Thanks to @sheldonhull over on the [Atuin warp issue](https://github.com/atuinsh/atuin/issues/1405): 

> FYI I just reload my profile in with exec zsh -l and then the nested version of that works fine. It's a work around but unblocked me.

if I `exec zsh` all my bindkeys work as expected - including Atuin. You could also just run a subshell and ignore the "warpify subshell" popup

**Comment by aviloff** [comment]
the `exec zsh` solution blocks all the warp features, but unblocked, for example, the ^j key combination that for me is vital when accepting lines. But this turned Warp into a simple terminal. I really wish this get handled soon! 👀

**Comment by GZTimeWalker** [comment]
Trying to migrate from alacritty, but I cannot do things like "Open VSCode in current dir" with `Command + E` hotkey.

As keybinds in zshrc not working, any solutions or alternatives for executing commands with some keys?

**Comment by nikitavoloboev** [comment]
It's genuinely insane this is not yet support by Warp. Can someone from dev team reply if this is even being looked at for support please? 🙏

**Comment by gpmcadam** [comment]
+1 for this feature. Like others, I'm trying to migrate my workflow from iTerm2 but Fish shell's `bind` functionality is core to my productivity. Simple things like binding custom Fish functions to keyboard shortcuts don't work in Warp, which means I'm forced to return to iTerm2 for now. The nested shell workaround isn't viable as it removes the very Warp features that make it attractive in the first place. Looking forward to proper shell keybinding support so I can give Warp another try. Good luck! :)

**Comment by varenc** [comment]
The lack of support here makes it feel like the target market for Warp is users who never wanted to personalize their key bindings.  Do Warp devs just not use bindkey? Would it be helpful to provide more examples of the many custom bindkey feature/workflows I use on a daily basis?  

It's sad for me to imagine die hard Warp users out there that never learn about the utility of custom zsh/fish function keybindings because they spend all their time in Warp where they're unsupported.  This is absolutely the #1 thing holding me back from using Warp.

**Comment by giuliobracci** [comment]
As others have said, this is really needed.

**Comment by Kristijan** [comment]
Adding another +1 to this request.

**Comment by WeiTing1991** [comment]
Same issue here with custom fzf keybinding. but it could run after `exec zsh`.
I'm trying to migrate my workflow from Wezterm to here

**Comment by Integralist** [comment]
I've decided to migrate to ghostty. I like Warp and have been using it a long time now but I needed bind key support.

**Comment by nikitavoloboev** [comment]
I don't get why it's hard to add. Perhaps it's cause they have to reimplement the bindkey support inside their `Warp` env?

There must be some way to hack this in, even if it's not performant or something.

**Comment by dreznicek** [comment]
This is a must have. For example, trying to use bindkey for using github copilot cli to auto explain or auto suggest with a zsh plugin... not joy. Pretty amazed warp has gone this long without this critical feature.

**Comment by KingMob** [comment]
Is this issue specifically about zsh's `bindkey`, or shell bindings in general? If it's not zsh-specific, the title should be clarified to something like "Support custom key bindings".

----

I noticed that Warpified bash doesn't support custom bindings, either for built-in readline fns or with custom `bind -x` code. Technically, the bindings DO get added to the keymaps (you can see with `bind -p` or `bind -X`), but the actual keyseqs are apparently never forwarded to the underlying bash to trigger them.

**Comment by kylesnowschwartz** [comment]
I think I've tried switching to Warp on 5 different occasions and find some annoyance that prevents me from doing so. This one is the latest.

**Comment by rrei** [comment]
Quite disappointing to see this feature being requested for so long, and by so many people, yet we still don't have it... 🫤

I'd love to switch to Warp, really, but having the default `fzf` `^R` keybinding is non-negotiable. So I'm forced to stay with `iTerm2` once again. I'm not saying iTerm is bad, but I'd like to try Warp for a while to decide which one to keep using, but you're making my decision very easy... 🫠

**Comment by Felipe-53** [comment]
Funniest thing, just discovered by accident that, if you run:

`exec zsh`

On a Warp terminal session, then the fzf shell binding starts working:

<img width="3456" height="1944" alt="Image" src="https://github.com/user-attachments/assets/357aaadb-1a23-484b-b277-9543059959b8" />

I suspect it would work for other keybindings defined in the zsh "environment"?

A dirty fix, but could maybe be used a first guide to the implementation?

**Comment by zg-alexey** [comment]
> Funniest thing, just discovered by accident that, if you run:
> 
> `exec zsh`

> A dirty fix, but could maybe be used a first guide to the implementation?

Yes, it is the simplest workaround. Also works with powershell or any other shell. Basically, you're creating a sub-shell, which disables Warp's integration completely, allowing to stock keybinds/integrations to work again.

**Comment by acarl005** [comment]
Not quite arbitrary bindkey support... but we're launching support for fzf and atuin to the [Warp Preview build](https://www.warp.dev/download-preview) tomorrow. Please try it out.

UPDATE: This is released on Preview version `v0.2026.09.02.08.27.preview_01`

**Comment by Felipe-53** [comment]
> Not quite arbitrary bindkey support... but we're launching support for fzf and atuin to the [Warp Preview build](https://www.warp.dev/download-preview) tomorrow. Please try it out.

Wow, that's huge. Thank you!
